DC-1111: Refactor SnapshotCreateFlight so SnapshotId is an input parameter

rjohanek opened the pull request:
## Summary

Ticket: https://broadworkbench.atlassian.net/browse/DC-1111

Before this PR, there was a step that created and stored the snapshotId in the working map. So, each step could access the snapshotId from the working map, but the snapshotId was not accessible in the flight. 

This PR moves the snapshotId creation into the service method that starts the flight. The snapshotId is now passed into the flight as an input parameter. This way it can be accessed directly in the flight instead of just in each step. So now, the snapshotId is passed into steps as a parameter instead of pulled from the working map.

## Motivation

The motivation for this change comes from this ticket: https://broadworkbench.atlassian.net/browse/DC-1028.
As part of this ticket which enables setting a DAC on snapshot create, I want to add two of the steps (CreateSnapshotGroupConstraintPolicyStep and AddSnapshotAuthDomainStep) from the SnapshotAddDataAccessControlsFlight to the SnapshotCreateFlight if a DAC is specified in the request model. The work in this PR is a prerequisite to that work because those steps take snapshotId as an input parameter. The other PR that is up for this ticket shows an old approach and does not yet represent this functionality. On that PR, this comment explains the change: https://github.com/DataBiosphere/jade-data-repo/pull/1693#discussion_r1613497126 

Additionally, creating the snapshot ID outside of the flight allows for more detailed logging which will help with debugging. By including snapshot ID in the description, the user themself is able to remove the locks remaining on the snapshot and dataset via TDR API calls, and delete the snapshot via TDR API call too. Instead of Support having to ask a TDR developer for the snapshot ID.

## Other Approaches

I considered passing in a null value for the snapshotId into CreateSnapshotGroupConstraintPolicyStep and AddSnapshotAuthDomainStep from the SnapshotCreateFlight. Then, updating those steps to include a null check and pulling the snapshotId from the working map if the snapshotId parameter is null. I avoided this approach to try to keep as much conditional logic out of steps as possible.

I considered modifying the SnapshotAddDataAccessControlsFlight to store the snapshotId in the working map. But this requires copying the value from the input parameters into the working map. This approach seemed less intuitive than the current implementation that just passes the snapshotId in as an input parameter, because the flight is updating an existing snapshot. 

I include these approaches to help document the thought process and open them up to feedback as well.

raejohanek committed:
move snapshotId to input parameter instead of working map

raejohanek committed:
sonarcloud fixes

pshapiro4broad commented on src/main/java/bio/terra/service/snapshot/SnapshotService.java:
```diff
@@ -184,12 +184,16 @@ public String createSnapshot(
       duosClient.getDataset(duosId, userReq);
     }
+
+    UUID snapshotId = UUID.randomUUID();
```
This is fine but you could inline it on line 196 too, I don't think a variable is needed here.

okotsopoulos replied:
@raejohanek I actually wanted to request that the flight description be expanded to include the UUID being passed in, in which case you'd define this UUID before the description for reuse there as well as an input parameter.  Ex.

```
UUID snapshotId = UUID.randomUUID();
String description = "Create snapshot %s with ID %s".formatted(snapshotRequestModel.getName(), snapshotId);
```

The reason for this is that this would make it easier for Support to assist users with dismally failed snapshot creation jobs.  Currently, Support needs to ask a TDR developer to go into the database to find the snapshot UUID that still needs to be cleaned up.  If this were exposed in the description, the user themself would be able to remove the locks remaining on the snapshot and dataset via TDR API calls, and delete the snapshot via TDR API call too.

But, this request is not directly related to your motivation, so no worries if you decline… I can add it in myself once this is in in its own ticket too.

rjohanek replied:
Yes I can add this change!

pshapiro4broad commented on src/main/java/bio/terra/service/snapshot/flight/create/SnapshotCreateFlight.java:
```diff
@@ -357,22 +370,24 @@ public SnapshotCreateFlight(FlightMap inputParameters, Object applicationContext
+          new CreateSnapshotCleanSynapseAzureStep(azureSynapsePdao, snapshotService, snapshotId));
     }
 
     addStep(new CreateSnapshotPolicyStep(policyService, sourceDataset.isSecureMonitoringEnabled()));
```
I think this step needs to be migrated too, see `CreateSnapshotPolicyStep`

pshapiro4broad commented on src/main/java/bio/terra/service/snapshot/flight/create/SnapshotCreateFlight.java:
```diff
@@ -113,6 +113,8 @@ public SnapshotCreateFlight(FlightMap inputParameters, Object applicationContext
     AuthenticatedUserRequest userReq =
         inputParameters.get(JobMapKeys.AUTH_USER_INFO.getKeyName(), AuthenticatedUserRequest.class);
 
+    UUID snapshotId = inputParameters.get(JobMapKeys.SNAPSHOT_ID.getKeyName(), UUID.class);
```
This is a breaking change with previous flights. To detect the breakage earlier, I would wrap this in a `Objects.requireNonNull()` so it fails before stairway tries to run a step. You can include a more specific error message for this case too, although that would only be useful for a short time after it's deployed, and then all new flights would have the input parameter set.

okotsopoulos replied:
Am I understanding correctly that this could only pose an issue during TDR deployments where snapshot creation flights were previously submitted, still running during deployment, and picked up by the new Stairway pod?

I agree that failing fast in those edge cases seems reasonable rather than putting a lot of effort to ensure a seamless transition.

pshapiro4broad replied:
> Am I understanding correctly that this could only pose an issue during TDR deployments where snapshot creation flights were previously submitted, still running during deployment, and picked up by the new Stairway pod?

Yes, exactly

pshapiro4broad reviewed (approved):
This looks good, but there's a more cleanup that should be done IMO.

To find the cleanups, I would look for every use of `SnapshotWorkingMapKeys.SNAPSHOT_ID` and remove it. For example, `UnlockSnapshotStep.doStep()` is still checking for this working map key.

There's some test cleanup that can be done too. For example, `CreateSnapshotSetResponseStepTest` is still creating a working map and stubbing `flightContext.getWorkingMap()`, but this step no longer uses the working map at all.

Also note that this change will break any snapshot create flight that's running when the new TDR is deployed, as it won't have a snapshot ID in its input parameters. I don't think this a huge cause for concern but it might be good to post in `#dsp-tdr` to let people know when it's coming out, or to ask if any large snapshot creations are in progress.

okotsopoulos commented on src/main/java/bio/terra/service/snapshot/flight/create/CreateSnapshotIdStep.java:
🎉

okotsopoulos commented on src/main/java/bio/terra/service/snapshot/flight/create/SnapshotCreateFlight.java:
```diff
@@ -357,22 +370,24 @@ public SnapshotCreateFlight(FlightMap inputParameters, Object applicationContext
 
     // unlock the resource metadata rows
-    addStep(new UnlockSnapshotStep(snapshotDao, null));
+    addStep(new UnlockSnapshotStep(snapshotDao, snapshotId));
```
I was looking at the source code for `UnlockSnapshotStep` and it looks like we also may be able to simplify it, as the special handling for null snapshot ID arguments was previously there only to support the snapshot creation case.

https://github.com/DataBiosphere/jade-data-repo/blob/19b780f5c80ab685fb27cbc2660cf478cdd3d63d/src/main/java/bio/terra/service/snapshot/flight/UnlockSnapshotStep.java#L38-L48

(But this also would be prone to the possibility of backwards incompatibility during deployments if included within this PR.)

rjohanek replied:
Yes, great idea I just simplified this and removed the tests for the removed cases!

okotsopoulos reviewed (approved):
I'm really excited for this change (see the Support use case in my comment) and could see us moving to this pattern for other resource creation too!

Happy to discuss the backwards incompatibility issues more as a group if desired.

fboulnois reviewed (commented):
Overall the change seems reasonable to me, I agree with the other concerns about handling breaking changes and including a log message for debugging.

raejohanek committed:
add snapshot ID to description

raejohanek committed:
require non-null snapshotId and update CreateSnapshotPolicyStep to take snapshotId as a parameter

raejohanek committed:
remove all usages of SnapshotWorkingMapKeys.SNAPSHOT_ID

rjohanek commented:
> This looks good, but there's a more cleanup that should be done IMO.
> 
> To find the cleanups, I would look for every use of `SnapshotWorkingMapKeys.SNAPSHOT_ID` and remove it. For example, `UnlockSnapshotStep.doStep()` is still checking for this working map key.

I removed all usages of the SNAPSHOT_ID key now!

> 
> There's some test cleanup that can be done too. For example, `CreateSnapshotSetResponseStepTest` is still creating a working map and stubbing `flightContext.getWorkingMap()`, but this step no longer uses the working map at all.
> 

I removed the use of the SNAPSHOT_ID key but CreateSnapshotSetResponseStep still needs to mock the working map because its puts the response and status code in the map.

> Also note that this change will break any snapshot create flight that's running when the new TDR is deployed, as it won't have a snapshot ID in its input parameters. I don't think this a huge cause for concern but it might be good to post in `#dsp-tdr` to let people know when it's coming out, or to ask if any large snapshot creations are in progress.

Okay, I can definitely post about this before merging it in to see when would be a good time for these changes to go into effect.

raejohanek committed:
add SNAPSHOT_ID input parameter to SnapshotCreateFlightTest
